Port to 1.21.4

modmuss50 opened the pull request:
#### TODO

- [x] Compile
- [x] Run
- [ ] Work

modmuss50 committed:
Bump versions to 1.21.4

ZurrTum committed:
DirectionProperty -> EnumProperty

Ayutac committed:
DirectionProperty -> EnumProperty

Ayutac committed:
FabricBlockSettings -> AbstractBlock.Settings

Ayutac committed:
Ported Nuke and related classes.

Ayutac committed:
Ported CableBlock.

Ayutac committed:
Ported PaddedShapedRecipeJsonBuilder.

Ayutac committed:
PacketCodecs.BOOL -> BOOLEAN.

Ayutac committed:
More PacketCodecs.BOOL -> BOOLEAN.

Ayutac committed:
getSize() -> size().

ZurrTum committed:
merge ItemActionResult and TypedActionResult into ActionResult

Ayutac committed:
Ported WrenchUtils.

ZurrTum committed:
PacketCodecs.BOOL -> BOOLEAN.

ZurrTum committed:
getBlockRegistryWrapper return RegistryEntryLookup<Block>

ZurrTum committed:
getTranslationKey() -> getItem().getTranslationKey()

ZurrTum committed:
Direction method rename

ZurrTum committed:
getBackgroundSprite return only one Identifier

ZurrTum committed:
getSlotType using DataComponent

ZurrTum committed:
Fluid isInfinite function parameter changes

ZurrTum committed:
slot canUseAsFuel need FuelRegistry

ZurrTum committed:
RebornExplosion @Override explode

ZurrTum commented:
I don't know how to submit here
https://github.com/ZurrTum/TechReborn/commits/port/1.21.4/

Ayutac commented:
You would have to make a PR to the branch. But maybe I can merge this.

modmuss50 commented:
> I don't know how to submit here https://github.com/ZurrTum/TechReborn/commits/port/1.21.4/

That looks good, you can make a PR to this branch here: https://github.com/TechReborn/TechReborn/compare/port/1.21.4...ZurrTum:TechReborn:port/1.21.4 (I got to this by going to the branch on your repo, and then clicking compare).

If you want I can add to you to this repo so you can push directly to this branch, it might make things a bit easier.

Ayutac commented:
I cherry-picked the changes and can commit them as well. Which method are we going for?

Ayutac committed:
Ported PaddedShapedRecipe.

Ayutac committed:
Ported RebornRecipe.

Ayutac committed:
Half-ported RecipeUtils.

Ayutac committed:
Ported RecipeUtils & SizedIngredient via UOE.

Ayutac committed:
Ported HologramRenderer.

Ayutac committed:
Ported AbstractConfigPopupElement.

Ayutac committed:
Removed out-of-date AWs.

Ayutac committed:
Ported GuiSprites.

Ayutac committed:
Ported GuiBase.

Ayutac committed:
Half-ported GuiBuilder.

Ayutac committed:
Ported rest of GuiBuilder.

Ayutac committed:
Hopefully ported ConfigSlotElement.

Ayutac committed:
TODO'd BlockOutlineRenderer.

Ayutac committed:
Ported ThemeManager.

ZurrTum commented:
How can I best contribute? Because sometimes modifications are repeated

ZurrTum commented:
BlockOutlineRenderer can replace: WorldRenderer.drawShapeOutline => DebugRenderer.drawVoxelShapeOutlines

Ayutac commented:
Hey if you want to continue, be my guest. This work sucks lol

ZurrTum committed:
Fix drawSpriteStretched call parameters

ZurrTum committed:
fix AbstractConfigPopupElement vertexConsumers

Ayutac commented:
We will need to rework the recipe system but I first want the code to compile and maybe, just maybe, get datagen to run

ZurrTum committed:
WorldRenderer -> DebugRenderer

ZurrTum commented:
I found two errors #3340

Ayutac committed:
Removed unneeded AW.

ZurrTum committed:
SimpleFramebuffer -> WindowFramebuffer

ZurrTum committed:
mixin getFovMultiplier adds two parameters

ZurrTum committed:
getSpriteAtlas using GuiAtlasManager

ZurrTum committed:
EntityAttributes remove GENERIC_

ZurrTum committed:
replace ArmorMaterial and ArmorItem.Type

ZurrTum committed:
Direction method rename

Ayutac committed:
LootContextParameterSet -> LootWorldContext

ZurrTum committed:
remove item canRepair function

ZurrTum committed:
remove wrapper

ZurrTum committed:
add miss parameter

Ayutac committed:
Fixed type generic.

ZurrTum committed:
getRegistryManager using getOrThrow

ZurrTum committed:
getItemBurnTime uses FuelRegistry

ZurrTum commented:
I need a short break, you can merge  #3342 first

Ayutac commented:
you are doing great. Thought you were asleep now xD

Ayutac committed:
Ported CloakingDeviceItem.

Ayutac committed:
Ported FuelRecipes.

Ayutac committed:
Ported TRToolMaterials. Not compatible with LevelZ anymore I'm afraid.

Ayutac committed:
Ported TRToolTier and got rid of TR(Tool)Item.

Ayutac committed:
Ported TREnergyArmourItem.

Ayutac committed:
Improved TREnergyArmourItem.

Ayutac committed:
Ported LightningRodBlockEntity.

Ayutac committed:
Ported FusionControlComputerBlockEntity.

Ayutac committed:
Ported StorageUnitBaseBlockEntity.

Ayutac committed:
Ported ElevatorBlockEntity.

Ayutac committed:
Removed gone isEnchantable.

ZurrTum committed:
fix GuiBuilder

ZurrTum commented:
> figure out getDisplays for RebornRecipe

By looking at the .class file I found that createIcon was replaced by DisplayItems, which was loaded from the json file
~~**Maybe it's safe to delete?**~~
### Path
createIcon <- displayItems <- addRecipes() <- show() <- RecipeBookAddS2CPacket.Entry <- sendInitRecipesPacket() <- RegistryKey<Recipe<?>> <- unlocked <- unlockRecipes() <- serverRecipeManager <- get() <- preparedRecipes <- PreparedRecipes.of() <- JsonDataLoader.load

ZurrTum commented:
Sorry, I ignored collector
```java
    @FunctionalInterface
    public interface DisplayCollector {
        void displaysForRecipe(RegistryKey<Recipe<?>> recipeKey, Consumer<RecipeDisplayEntry> adder);
    }
```
createIcon() replaced with getDisplays(), Requires **Override getDisplays function**:
```java
    private static List<ServerRecipe> collectServerRecipes(Iterable<RecipeEntry<?>> recipes, FeatureSet enabledFeatures) {
         ...
        for(RecipeEntry<?> recipeEntry : recipes) {
            Recipe<?> recipe = recipeEntry.value();
            ...
            for(RecipeDisplay recipeDisplay : recipe.getDisplays()) {
                if (recipeDisplay.isEnabled(enabledFeatures)) {
                    int i = list.size();
                    NetworkRecipeId networkRecipeId = new NetworkRecipeId(i);
                    RecipeDisplayEntry recipeDisplayEntry = new RecipeDisplayEntry(networkRecipeId, recipeDisplay, optionalInt, recipe.getRecipeBookCategory(), optional);
                    list.add(new ServerRecipe(recipeDisplayEntry, recipeEntry));
                }
            }
        }

        return list;
    }
```

ZurrTum committed:
FabricEntityTypeBuilder -> EntityType.Builder

ZurrTum committed:
createFuelTimeMap -> FuelRegistry

ZurrTum committed:
hasRecipeRemainder -> getRecipeRemainder

ZurrTum committed:
world.getRecipeManager -> server.getRecipeManager

Ayutac committed:
Ported ElectricFurnaceBlockEntity.

Ayutac committed:
Ported PaintingToolItem.

Ayutac committed:
Ported RockCutterItem.

Ayutac committed:
Ported IronFurnaceBlockEntity.

Ayutac committed:
Ported AutoCraftingTableBlockEntity.

Ayutac committed:
Half-ported RollingMachine with UOE.

ZurrTum committed:
implementing getDisplays

Ayutac committed:
Ported rest of RollingMachine with another UOE.

Ayutac committed:
Ported ResinBasinBlock.

ZurrTum commented:
#3345 implementing getDisplays
I need to rest. I'll continue tomorrow.

Ayutac committed:
Ported TRRecipeHandler via UOE.

Ayutac committed:
Ported Tools.

Ayutac committed:
Ported last remaining Tools.

Ayutac committed:
Ported Recipe-related Groovy.

Ayutac committed:
Half-ported Groovy ModelProvider.

Ayutac committed:
Half-ported Groovy MixinTextureMap.

ZurrTum committed:
implementing unlockTRRecipes

Ayutac committed:
Exchanged reflection with AW.

ZurrTum committed:
DefaultCraftingDisplay<?> -> DefaultCraftingDisplay

ZurrTum committed:
Ported NukeRenderer

ZurrTum committed:
Direction asRotation rename getPositiveHorizontalDegrees

ZurrTum committed:
ModelTransformationMode transferred to net.minecraft.item

ZurrTum committed:
drawContext.drawTexture add parameters

ZurrTum committed:
drawContext.drawSprite -> drawSpriteStretched

ZurrTum committed:
split ReiPlugin to implement REICommonPlugin

ZurrTum committed:
implements REI Display

ZurrTum committed:
remove redundant imports

ZurrTum committed:
Ported GuiAutoCrafting

ZurrTum committed:
BLOCK_ATLAS_TEXTURE transferred to net.minecraft.client

ZurrTum committed:
Ported TurbineRenderer

ZurrTum committed:
RebornExplosion getAffectedBlocks rename

ZurrTum committed:
JsonUnbakedModel.getTransformations to static UnbakedModel

ZurrTum committed:
remove unused FurnaceFuelSlot fix canUseAsFuel

ZurrTum commented:
Is it because of the beta version that Fabric-api 1.21.4-pre3 does not have Fabric-model-loading-api-v1-0.110.0.jar?
Do I need to implement a mixin?

Ayutac commented:
I think it is best to wait for that. @modmuss50 ?

modmuss50 commented:
> I think it is best to wait for that. @modmuss50 ?

Yes, wait on this, its pending a FAPI release, should be really soon.

This looks like an awesome amount of work 👍 Ive been busy working on Fabric stuff for 1.21.4 so will have to catch up here soon.

ZurrTum commented:
Hey, can you open port/1.21.3 first?
https://github.com/ZurrTum/TechReborn/commits/port/1.21.3/

ZurrTum commented:
Need to add the port/1.21.3 branch at the current location

ZurrTum commented:
## `1.21.3`
### gradlew build
### $${\color{green}BUILD \space SUCCESSFUL}$$

### Run minecraft
### $${\color{green}OK}$$

### Unable to join world:
Couldn't parse data file 'techreborn:blast_furnace/tungsten_ingot' from 'techreborn:recipe/blast_furnace/tungsten_ingot.json': DataResult.Error['Failed to parse either. First: Not a JSON object: "#c:ores/raw_tungsten"; Second: Missing tag: 'c:ores/raw_tungsten' in 'minecraft:item'': BlastFurnaceRecipe[type=techreborn:blast_furnace, ingredients=[], outputs=[1 techreborn:tungsten_ingot], power=128, time=2000, heat=2500]]

modmuss50 committed:
Misc fixes, some todo pending on FAPI changes

modmuss50 committed:
Main menu

modmuss50 committed:
Merge remote-tracking branch 'origin/1.21' into port/1.21.4

# Conflicts:
#	RebornCore/src/main/java/reborncore/common/blockentity/SlotConfiguration.java
#	RebornCore/src/main/java/reborncore/common/network/serverbound/IoSavePayload.java

modmuss50 committed:
Fix after merge

modmuss50 commented:
> Hey, can you open port/1.21.3 first? https://github.com/ZurrTum/TechReborn/commits/port/1.21.3/

Im not sure how useful 1.21.3 will be, I think we are mostly at the stage where the issues are the same between both versions. I have got 1.21.4 to the main menu, the next thing to do is get datagen to run.

ZurrTum commented:
Ok I will check datagen for 1.21.4 and migrate from port/1.21.3 to port/1.21.4

ZurrTum commented:
@modmuss50 There is a problem with the new gradle plugin, how to set it up?
Execution failed for task ':processResources'.
> Entry .cache/07fec851283d32f404434541afae212659cc2b69 is a duplicate but no duplicate handling strategy has been set.

modmuss50 commented:
> @modmuss50 There is a problem with the new gradle plugin, how to set it up? Execution failed for task ':processResources'.
> 
> > Entry .cache/07fec851283d32f404434541afae212659cc2b69 is a duplicate but no duplicate handling strategy has been set.

I had to delete the /src/generated folder, I'm not 100% why that's happening, once datagen is running it will be easier to look at.

ZurrTum commented:
#3351
see action https://github.com/TechReborn/TechReborn/actions/runs/12106084831/job/33751449613?pr=3351
Execution failed for task ':processResources'.
> Entry data/minecraft/tags/block/mineable/hoe.json is a duplicate but no duplicate handling strategy has been set. Please refer to https://docs.gradle.org/8.11.1/dsl/org.gradle.api.tasks.Copy.html#org.gradle.api.tasks.Copy:duplicatesStrategy for details.

ZurrTum commented:
@modmuss50 Port/1.21.4 #3351 is Complete 😄

ZurrTum committed:
Port/1.21.4 (#3351)

* fix datagen TechRebornRecipesProvider extends FabricRecipeProvider

* fix datagen Ingredient.fromTag

* fix datagen RecipesProvider call RecipeGenerator

* fix datagen Builder.create

* fix datagen Identifier -> RegistryKey<Recipe>

* fix datagen Ingredient codec

* fix datagen ofStacks -> fromTag

* fix datagen itemLookup

* fix datagen Identifier -> RegistryKey<Recipe>

* TODO ItemTags.TRIM_TEMPLATES

* Apply spotless

* fix gradle duplicate handling strategy

modmuss50 committed:
Fix duplicates

modmuss50 commented:
> @modmuss50 Port/1.21.4 #3351 is Complete 😄

Thanks :) Merged and pushed a fix for the duplicates

ZurrTum commented:
Now we need to solve the problem of Codec serialization Ingredient.CODEC
> Couldn't parse data file 'techreborn:blast_furnace/tungsten_ingot' from 'techreborn:recipe/blast_furnace/tungsten_ingot.json': DataResult.Error['Failed to parse either. First: Not a JSON object: "#c:ores/raw_tungsten"; Second: Missing tag: 'c:ores/raw_tungsten' in 'minecraft:item'': BlastFurnaceRecipe[type=techreborn:blast_furnace, ingredients=[], outputs=[1 techreborn:tungsten_ingot], power=128, time=2000, heat=2500]]

I can't find the entrance, can you help me?

Ayutac commented:
Caused by: java.lang.RuntimeException: Cannot load class net.fabricmc.fabric.api.client.datagen.v1.provider.FabricModelProvider in environment type SERVER

modmuss50 committed:
Make it not crash enough to grind some ore

modmuss50 committed:
Fix translation keys

modmuss50 committed:
Fix smelting recipes

modmuss50 committed:
Use a script to fix most crafting recipes

modmuss50 commented:
> Caused by: java.lang.RuntimeException: Cannot load class net.fabricmc.fabric.api.client.datagen.v1.provider.FabricModelProvider in environment type SERVER

You will likely need to re-create your run configuration for the datagen as it now runs on the client.

ZurrTum commented:
@modmuss50 In the process of fixing the recipe, I found a bug
https://github.com/TechReborn/TechReborn/commit/061369a914697d0ae352ac85002b3922f96b0478#diff-7c7779ce723b20917a8e1c5fe2bfaedb9a60a85e550ec3cadf6b9e1014f93161L13
Before renaming it was `endstone_small_dusts`, after renaming it was `dusts/endstone_small`
https://github.com/TechReborn/TechReborn/blob/0bbed0fe908f58369ac605a51c1436a3bb748edf/src/main/resources/data/techreborn/recipe/crafting_table/parts/shulker_shell.json#L13
But the generated label is `small_dusts/endstone`
https://github.com/TechReborn/TechReborn/blob/0bbed0fe908f58369ac605a51c1436a3bb748edf/src/main/java/techreborn/init/TRContent.java#L931
and there are errors elsewhere:
https://github.com/TechReborn/TechReborn/blob/0bbed0fe908f58369ac605a51c1436a3bb748edf/src/datagen/groovy/techreborn/datagen/TRConventionalTags.groovy#L50

## Need to check
|              before               |                   after                   |
| :-------------------------------: | :---------------------------------------: |
| c:iridium_reinforced_stone_blocks | c:storage_blocks/iridium_reinforced_stone |
|          c:steel_blocks           |          c:storage_blocks/steel           |
|      c:tungstensteel_blocks       |      c:storage_blocks/tungstensteel       |
| c:iridium_reinforced_stone_blocks | c:storage_blocks/iridium_reinforced_stone |

|   Missing   |
| :---------: |
| c:limestone |
|  c:marble   |
|  c:sulfurs  |

more see #3352, maybe you need to modify it again

ZurrTum committed:
fix parse data file Missing tag (#3352)

ZurrTum committed:
Fix block model texture key

down -> bottom
fishing_station.json add missing key

ZurrTum committed:
nanosaber.json add missing key

ZurrTum committed:
Add items folder to set up the model

ZurrTum commented:
@modmuss50 By looking at the minecraft resources folder I found that the items folder is missing
By default, only block item configurations will be automatically generated.
```json5
// items/{{block_name}}.json
{
  "model": {
    "type": "minecraft:model",
    "model": "techreborn:block/{{block_name}}"
  }
}
```

### Block
| default | need |
| :-: | :-: |
| techreborn:block/basic_storage_unit | techreborn:block/machines/tier0_machines/basic_storage_unit |

### Item
| default | need |
| :-: | :-: |
| null | techreborn:item/{{item_name}} |

#3355

ZurrTum commented:
I knew I might need datagen to generate it, so I opened another branch

ZurrTum committed:
Use ItemModel to implement dynamic fluids

ZurrTum commented:
#3357 implement dynamic fluids 🎉

ZurrTum committed:
Fix missing toast icon for scrapbox

ZurrTum committed:
fix datagen add ScrapboxRecipeJsonFactory

ZurrTum committed:
Remove unused imports

ZurrTum commented:
The REI plugin doesn't support 1.21.4 yet, how do I test recipe related stuff?

ZurrTum committed:
Fix some REI error

ZurrTum committed:
fix RollingMachine UOE

ZurrTum committed:
Fix power bar energy display

Cancel the scissor effect of power_bar_overlay

ZurrTum commented:
@Ayutac Do you have time to merge the pull request for me?
Some json files I don't know if I need them, maybe I can finish the datagen today.
Should I just issue a single pull request with multiple commits?

Ayutac commented:
done

ZurrTum committed:
Fix AutoCraftingTable output preview

Synchronized recipe output since only the server now knows all recipes

ZurrTum committed:
Fix IronFurnace client cooking time

ZurrTum committed:
Fix IronAlloyFurnace client cooking time

ZurrTum committed:
Do not call canSmelt repeatedly

ZurrTum committed:
Fix FluidReplicatorRecipe onCraft add amount

ZurrTum commented:
@Ayutac Contains some fixes and you can merge them. Also includes the 1.21 branch.
I will try to write datagen for ItemModel later, which should take a while

ZurrTum committed:
Add datagen texture path

ZurrTum committed:
Add data generation for all models

ZurrTum committed:
Make texture image names more suitable for data generation

ZurrTum committed:
Delete blockstates and items and models folders that are no longer needed

ZurrTum commented:
@modmuss50 #3366 spent me a lot of time. I would like to know when 1.21.4 is planned to enter the beta version, and I will continue to contribute code recently.

modmuss50 commented:
> @modmuss50 #3366 spent me a lot of time. I would like to know when 1.21.4 is planned to enter the beta version, and I will continue to contribute code recently.

Sorry I havent been keeping up, if you think its ready for a beta I am happy to merge this PR and release it as a beta.

ZurrTum commented:
Tomorrow I will test all items, maybe tomorrow

ZurrTum committed:
Fix armor slot

add Image to gui atlas

ZurrTum committed:
Fix FusionControlComputer attacking players

ZurrTum committed:
Fix IronFurnace experience tooltip

ZurrTum committed:
Fix nuke

ZurrTum committed:
Add ChunkLoader translate

ZurrTum committed:
Add Launchpad translate

ZurrTum committed:
Fix \t character errors

ZurrTum commented:
@modmuss50 #3367
I think it's time to release the beta version

ZurrTum commented:
The REI plugin has not been integrated yet. If you have time, you can also deal with it.

Ayutac commented:
It appears that the recipe advancements are on by default, this should be corrected.

Ayutac committed:
Added some 1.21.4 recipes.

Pale Oak in Sawmill, eyeblossom in extractor, sap <-> resin in crafting table.

ZurrTum commented:
What more work needs to be done?
I will make time to deal with it tomorrow

ZurrTum commented:
I'm not familiar with the REI plug-in. After trying to open it, there are only 8 recipes. I don't know how to solve it. Maybe I need to spend time looking at the code.

ZurrTum commented:
@modmuss50 If you are really too busy, you can give me more permissions to handle it. But I can only ensure that I have free time to deal with it in the next few months, and there is no guarantee for a long time.

modmuss50 commented:
Im happy to take a look, might be later this week. Im going to have some free time over the next few weeks so I can finish this off.

What you have done already is a massive help.

ZurrTum commented:
I'm happy to be able to help with this project and it's how I spend my time

ZurrTum committed:
Fix REI plugin recipe display

ZurrTum commented:
#3369
By turning on the REI debug mode, a recipe serialization error was found, which has been fixed.
The REI 18 version is not yet stable, and some recipes of the built-in plug-ins are also wrong. Currently, there are still some minor problems with dynamic fluid units and fluids in REI. The reasons are not yet known.

Ayutac committed:
Generalized the tuff centrifuge recipe.

Now also accepts chiseled tuff, tuff bricks, polished tuff and chiseled tuff bricks, as well as anything in c:tuff tag.

ZurrTum commented on src/datagen/groovy/techreborn/datagen/recipes/machine/centrifuge/CentrifugeRecipesProvider.groovy:
```diff
@@ -509,7 +509,7 @@ class CentrifugeRecipesProvider extends TechRebornRecipesProvider {
 			power 5
 			time 2500
-			ingredients stack(Items.TUFF, 16)
+			ingredients stack(TRConventionalTags.TUFF, 16)
```
I see the build failed, maybe it can be changed to tag call:
```
ingredient {
	tag(TRConventionalTags.TUFF, 16)
}
```

Ayutac replied:
You are too fast, I have corrected it locally already, give me a minute :D

Ayutac committed:
Fixed bug introduced in previous commit.

ZurrTum commented:
@Ayutac There was previously an open pull request #3369, you can merge it

Ayutac committed:
Made copper oxidizable in electrolyser.

Ayutac committed:
Replaced workbench with crafter in autocrafter recipe.

ZurrTum committed:
Fix REI slot display ComponentsIngredient

minecraft:stone -> techreborn:cell

ZurrTum committed:
Unified TemplateState code style

ZurrTum committed:
Fix fusion control computer recipe energy consumed

ZurrTum commented:
@Ayutac I found a bug in the fusion control computer recipe and need you to see if it can be fixed. #3371
Also, some questions maybe you can close it: #3368 #3364 #3350 #3310 #3294

modmuss50 reviewed (commented):
Looks great, thanks both for the help. I will release this as a beta.

modmuss50 committed:
Merge branch '1.21' into port/1.21.4

ZurrTum commented:
Glad it has reached beta version.
@Ayutac Do the Tests and TODOs in the txt file below docs need to be resolved?
In terms of testing, I have already tested some. I am not sure whether TODO is necessary.

Ayutac commented:
Need is a strong word. Can definitely be released without being resolved
